File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs, map = new Map()) {
    
    if (!strs) return [];
    groupWords(strs, map)
    
    return [...map.values()];
};



const groupWords = (words, map) => {
    for (const original of words) {
        const sorted = reorder(original);
        const values = map.get(sorted) || [];
        
        values.push(original);
        map.set(sorted, values);
    }
}

const reorder = str => str.split('').sort((a,b) => a.localeCompare(b)).join('');

File: 0049-group-anagrams/0049-group-anagrams.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """

        hashmap = {}
        values = []
        for original in strs:
            orderedStrs = ''.join(sorted(original))
            values = hashmap.get(orderedStrs) or []

            values.append(original)
            hashmap[orderedStrs] = values
        return list(hashmap.values()) 
        